File: cli/qc.mjs
#!/usr/bin/env node
/**
 * BrowserQC CLI — run the real app headlessly and write the QC metrics as JSON.
 *
 *   npm run build                                    # once, produces dist/
 *   node cli/qc.mjs --in T1.nii.gz --out results.json
 *
 * This drives the actual BrowserQC page in headless Chrome rather than
 * re-implementing the pipeline in Node, so the numbers are the browser's by
 * construction. That is not just convenience: segmentation needs WebGPU, which Node
 * has no implementation of. The wasm module's only backend is `webgpu` and it
 * refuses rather than silently falling back to its (validation-only, minutes-per-
 * volume) CPU engine.
 *
 * The input is injected by intercepting the page's request for its bundled default
 * image, so the app's normal auto-run does all the work and no app code is special-
 * cased for the CLI. The one seam is `window.browserqcMetrics` (see computeQc in
 * src/main.ts), which exposes full-precision values the panel would round to 3 s.f.
 */
import { chromium } from 'playwright'
import { spawn } from 'node:child_process'
import { readFile, writeFile, access } from 'node:fs/promises'
import { readFileSync } from 'node:fs'
import { basename, dirname, join, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), '..')
// Deploy base from vite.config, so the CLI targets the right path whether the app is
// served at a /repo/ subpath or a custom-domain root (base: '/').
const BASE = readFileSync(join(ROOT, 'vite.config.ts'), 'utf8').match(/base:\s*'([^']*)'/)?.[1] ?? '/'
const log = (m) => process.stderr.write(`${m}\n`)

function parseArgs(argv) {
  const a = { in: null, out: null, bids: null, port: 4199, timeout: 300000 }
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === '--in') a.in = argv[++i]
    else if (argv[i] === '--out') a.out = argv[++i]
    else if (argv[i] === '--bids') a.bids = argv[++i]
    else if (argv[i] === '--port') a.port = Number(argv[++i])
    else if (argv[i] === '--timeout') a.timeout = Number(argv[++i]) * 1000
    else if (argv[i] === '--help' || argv[i] === '-h') a.help = true
  }
  return a
}

const args = parseArgs(process.argv.slice(2))
if (args.help || !args.in || !args.out) {
  process.stdout.write(
    'Usage: node cli/qc.mjs --in <T1.nii[.gz]> --out <results.json> [--bids sidecar.json]\n' +
      '                       [--port N] [--timeout SEC]\n' +
      '\nRequires a production build (npm run build) and Google Chrome.\n',
  )
  process.exit(args.help ? 0 : 1)
}

try {
  await access(join(ROOT, 'dist', 'index.html'))
} catch {
  log('error      no dist/ — run `npm run build` first')
  process.exit(1)
}

const inputBytes = await readFile(args.in)
const URL_BASE = `http://localhost:${args.port}${BASE}`

// --- boot vite preview on the production build --------------------------
// detached so we can kill the whole process group (vite + esbuild children).
const preview = spawn('npx', ['vite', 'preview', '--port', String(args.port), '--strictPort'], {
  cwd: ROOT,
  stdio: 'ignore',
  detached: true,
})
let previewExited = false
preview.on('exit', () => { previewExited = true })
const killPreview = () => { try { process.kill(-preview.pid, 'SIGTERM') } catch { /* gone */ } }
process.on('exit', killPreview)
process.on('SIGINT', () => { killPreview(); process.exit(130) })

const wait = (ms) => new Promise((r) => setTimeout(r, ms))
async function waitForServer(deadlineMs = 20000) {
  const until = Date.now() + deadlineMs
  while (Date.now() < until) {
    if (previewExited) throw new Error(`vite preview exited — is port ${args.port} in use?`)
    try { if ((await fetch(URL_BASE)).ok) return } catch { /* not up yet */ }
    await wait(250)
  }
  throw new Error('vite preview did not come up')
}

let browser
try {
  await waitForServer()
  log(`input      ${basename(args.in)}`)

  // System Chrome for full WebGPU (NiiVue needs it). The angle/swiftshader flags are
  // the same software-rendering fallback the e2e smoke uses, so this works on a
  // headless machine with no GPU; on a real GPU Chrome uses it and runs faster.
  browser = await chromium.launch({
    headless: true,
    channel: 'chrome',
    args: ['--use-gl=angle', '--enable-unsafe-swiftshader', '--window-size=1280,960'],
  })
  const page = await browser.newPage()

  const pageErrors = []
  page.on('pageerror', (e) => pageErrors.push(e.message))
  page.on('console', (m) => { if (m.type() === 'error') pageErrors.push(m.text()) })

  // Serve OUR file wherever the app asks for its bundled default image, so the
  // standard startup auto-run (segment → overlay → --qc) processes it.
  await page.route('**/t1_crop.nii.gz', (route) =>
    route.fulfill({ status: 200, contentType: 'application/gzip', body: inputBytes }),
  )
  // The injected image is not the default subject, so suppress the default subject's
  // BIDS sidecar (resolve it to JSON null — a 404 would trip the error gate below).
  // --bids (below) supplies its own via window.__browserqcBids.
  await page.route('**/t1_crop.json', (route) =>
    route.fulfill({ status: 200, contentType: 'application/json', body: 'null' }),
  )

  if (args.bids) {
    // Same seam the drag-drop sidecar uses: metadata only, never affects metrics.
    const meta = JSON.parse(await readFile(args.bids, 'utf8'))
    await page.addInitScript((m) => { window.__browserqcBids = m }, meta)
  }
  await page.goto(URL_BASE, { waitUntil: 'domcontentloaded' })
  log('running    segmentation → niimath --qc (headless Chrome)…')

  // NB: waitForFunction(fn, arg, options) — the options object must be the THIRD
  // argument; passing it second silently falls back to Playwright's 30 s default.
  await page.waitForFunction(() => window.browserqcMetrics !== undefined, null, {
    timeout: args.timeout,
  })
  const report = await page.evaluate(() => window.browserqcMetrics)
  const status = await page.$eval('#statusMsg', (el) => el.textContent || '').catch(() => '')

  // The app is expected to run clean (the smoke gates on the same signal); a page/console
  // error means the result may be wrong, so fail rather than bless it.
  if (pageErrors.length) throw new Error(`page reported ${pageErrors.length} error(s):\n  ${pageErrors.join('\n  ')}`)

  await writeFile(args.out, `${JSON.stringify(report, null, 2)}\n`)
  log(`done       ${args.out}  (${status.trim()})`)
} catch (err) {
  log(`error      ${err.message}`)
  process.exitCode = 1
} finally {
  if (browser) await browser.close().catch(() => {})
  killPreview()
}

File: cli/qc.py
#!/usr/bin/env python3
"""BrowserQC without a browser: the page's pipeline on the native executables, which
must be on PATH (or in $BROWSERQC_BIN). Standard library only, Python 3.8+.
  brainchop-<model>  https://github.com/neuroneural/brainchopC/releases
  niimath            https://pypi.org/project/niimath/  (--qc --pve needs v1.0.20260926+)

  python3 cli/qc.py --in T1.nii[.gz] --out qc.json [--model 16chan18cls|mindmap|mindsnap|mindmap-pve] [--bids sidecar.json]
"""
import argparse
import json
import os
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
MODELS = json.loads((ROOT / "src" / "models.json").read_text())
TEMPLATE = ROOT / "public" / "avg152T1.nii.gz"


def run(exe, *args, path):
    found = shutil.which(exe, path=path)
    if not found:
        raise RuntimeError(f"{exe} not found: put it on PATH or in $BROWSERQC_BIN")
    if subprocess.run([found, *map(str, args)], stdout=subprocess.DEVNULL).returncode:
        raise RuntimeError(f"{exe} failed")


def main():
    parser = argparse.ArgumentParser(description="MRIQC-style QC of a T1 image with brainchop + niimath.")
    parser.add_argument("--in", dest="input", required=True, help="T1.nii[.gz]")
    parser.add_argument("--out", required=True, help="report JSON")
    parser.add_argument("--model", default="16chan18cls", choices=list(MODELS))
    parser.add_argument("--bids", help="BIDS sidecar to embed as bids_meta")
    args = parser.parse_args()
    path = os.pathsep.join(filter(None, [os.environ.get("BROWSERQC_BIN"), os.environ.get("PATH")]))
    model = MODELS[args.model]
    t1 = os.path.abspath(args.input)  # a leading '-' must not read as an option
    with tempfile.TemporaryDirectory(prefix="browserqc-") as tmp:
        tmp = Path(tmp)
        if model.get("pve"):
            run(f"brainchop-{model['pve']}", t1, "--pve", "-o", tmp / "pve.nii", path=path)  # writes pve_{gm,wm,csf}.nii
            tissues = ["--pve", *(tmp / f"pve_{t}.nii" for t in ("csf", "gm", "wm"))]
        else:
            run(f"brainchop-{args.model}", t1, "-o", tmp / "seg.nii", path=path)
            tissues = ["--seg", tmp / "seg.nii", "--csf", ",".join(map(str, model["csf"])),
                       "--wm", ",".join(map(str, model["wm"]))]
        run("niimath", "--qc", t1, *tissues, "--air", TEMPLATE, "--json", tmp / "qc.json", path=path)
        report = json.loads((tmp / "qc.json").read_text())
    # Match the page's report: its provenance and bids_meta, and the template's bare name.
    report["provenance"]["air_template"] = TEMPLATE.name
    report["provenance"]["segmentation"] = f"brainchop {args.model} ({model['label']})"
    if args.bids:
        report["bids_meta"] = json.loads(Path(args.bids).read_text())
    Path(args.out).write_text(json.dumps(report, indent=2, ensure_ascii=False) + "\n")


if __name__ == "__main__":
    try:
        main()
    except (RuntimeError, OSError, ValueError) as err:  # ValueError: a malformed --bids sidecar
        sys.exit(f"error: {err}")
